aws-api-appsync/src/main/java/com/amplifyframework/datastore/appsync/ModelWithMetadataAdapter.java

```diff
@@ -60,7 +61,15 @@ public static void register(@NonNull GsonBuilder builder) {
             throw new JsonParseException("Expected a parameterized type during ModelWithMetadata deserialization.");
         }
 
-        Model model = context.deserialize(json, modelClassType);
+        final Model model;
```

Comment:
Hmm, why is this change necessary?  Shouldn't the `SerializedModelAdapter` already handle this?

If this is needed here, then I think we would probably also need a similar change in any other adapter that deserializes a Class that could contain a Model (e.g. `PaginatedResult`, `Iterable`, `GraphQLResponse`)

Reply:
Alright, I studied this change a bit more. I think, for now at least, it probably makes sense to leave it as is.

The `SerializedModelAdapter` currently stores `modelSchema` and `serializedData`,into a JSON structure like this:
```json
{
    "modelSchema": {
        ...
    },
    "serializedData": {
    }
}
```

However, the format of the `ModelWithMetadata` is like:
```json
{
    "flatDataKey1": "dataVal",
    "flatDataKey2": "dataVal",
    "_metadataStuff1": "metadata",
    "_metadataStuff2": "metadata"
}
```

The GSON deserialization will cascade _through_ the adapters you mention, and will try to deserialize a `GraphQLResponse<Iterable<ModelWithMetadata<SerializedModel>>>`. It knows how to do all of that just fine, except when it gets to the inner-most adapter. At that point, it looks for a field called `serializedData` in that second JSON structure (which doesn't exist.)

So I think TLDR: I _am_ actually doing some custom deserialization only for `ModelWithMetadata`.

And, it isn't needed for other adapters, since DataStore will _always_ envelope in the `ModelWithMetadata`.

--------

I can think of a few others ways we could do this.

1. At a higher level, pass the `T` as `Map<String, Object>`, s.t. we try to deserialize `GraphQLResponse<Iterable<ModelWithMetadata<Map<String, Object>>>>`. Then, we could rectify the map data back into a `SerializedModel.builder().serializedData(theResponseContent).build()` at a higher level in the stack.

2. Rip `<T>` out of almost all of the system, in future work. This is what I _actually_ want to do. With this, the SQL layer and the AppSync sync layer will just know about `SerializedModel` only, but not any model types. (I also propose that we rename `SerializedModel` to maybe just `Blob`, but that's a separate matter.) Only the highest layer of the system (just under the DataStore API) would know about Java-language data-modeling. At that high-level, it would quickly convert to schema/blob object pairs. Sync engine & storage adapter would operate wholly on those.

Reply:
Got it, thanks for investigating it :). I'm good to leave it as is for now.

> Rip <T> out of almost all of the system, in future work.

I agree this is the right direction to move towards.

> I also propose that we rename SerializedModel to maybe just Blob, but that's a separate matter.

:+1:

I wonder if we could also eliminate `ModelWithMetadata` entirely, and just store the metadata as a part of the `Blob`.  It would be closer to how the data is modeled on the AppSync side anyway.   Seems like this could simplify a lot too.